### PR TITLE
Query model permissions

### PR DESCRIPTION
**Pull request by thibaultrobert**
Works with https://github.com/SubstraFoundation/substra-backend/pull/117

### CODE REVIEW

**Comment by jmorel** [comment]
I tested this PR against https://github.com/SubstraFoundation/substra-backend/pull/117/files and it works like a charm!

File: README.md [hunk 1]
```diff
@@ -34,6 +34,7 @@ Note for internal use only: See the [technical specifications](https://github.co
 
 ### Implemented smart contracts
 
+- `cancelComputePlan`
```
Comment (inelgnu) [inline README.md:37]: is this related to this PR?

Comment (thibaultrobert) [reply README.md:37]: Yes and no: its because I updated this list with the missing smart contract.

File: chaincode/tuple.go [hunk 2]
```diff
@@ -256,3 +294,11 @@ func determineTupleStatus(db *LedgerDB, tupleStatus, computePlanID string) (stri
...
 	return tupleStatus, nil
 }
+
+func createModelIndex(db *LedgerDB, tupleKey, modelHash string) error {
+	err := db.CreateIndex("tuple~modelHash~key", []string{"tuple", modelHash, tupleKey})
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (AurelienGasser) [inline chaincode/tuple.go]: unused

File: chaincode/tuple.go [hunk 3]
```diff
@@ -163,6 +163,44 @@ func queryModels(db *LedgerDB, args []string) (outModels []outputModel, err erro
 	return
 }
 
+func queryModelPermissions(db *LedgerDB, args []string) (outputPermissions, error) {
```
Comment (AurelienGasser) [inline chaincode/tuple.go:166]: It took me a long time to understand this function. I think more explicit naming would help make it more clear (see comments below)

File: chaincode/tuple.go [hunk 4]
```diff
@@ -163,6 +163,44 @@ func queryModels(db *LedgerDB, args []string) (outModels []outputModel, err erro
...
+	var out outputPermissions
+	inp := inputKey{}
+	err := AssetFromJSON(args, &inp)
+	if err != nil {
+		return out, err
+	}
```
Comment (AurelienGasser) [inline chaincode/tuple.go:172]: Here for instance, add:

```go
modelHash := inp.Key
```

and use `modelHash` throughout the function instead of `inp.Key`

File: chaincode/tuple.go [hunk 5]
```diff
@@ -163,6 +163,44 @@ func queryModels(db *LedgerDB, args []string) (outModels []outputModel, err erro
...
+	keys, err := db.GetIndexKeys("tuple~modelHash~key", []string{"tuple", inp.Key})
+	if err != nil {
+		return out, err
+	}
+	if len(keys) == 0 {
+		return out, errors.NotFound("Could not find a model for hash %s", inp.Key)
+	}
+	key := keys[0]
```
Comment (AurelienGasser) [inline chaincode/tuple.go]: Here too:

```suggestion
	tupleKey := keys[0]
```

File: chaincode/tuple.go [hunk 6]
```diff
@@ -163,6 +163,44 @@ func queryModels(db *LedgerDB, args []string) (outModels []outputModel, err erro
...
+	modelPermissions := Permissions{}
+	modelPermissions.Process.Public = true
+
+	// get out-model and permissions from parent for head model in composite traintuple
+	if tupleType == CompositeTraintupleType{
+		tuple, err := db.GetCompositeTraintuple(key)
+		if err != nil {
+			return out, errors.Internal(err, "queryModelPermissions:")
+		}
+		if tuple.OutHeadModel.OutModel.Hash == inp.Key {
```
Comment (AurelienGasser) [inline chaincode/tuple.go]: Maybe add a comment here? Something like 

```go
// if `modelHash` refers to the head out-model, return the head out-model permissions
// if `modelHash` refers to the trunk out-model, do nothing (default to "public processable")
```

File: chaincode/tuple_test.go [hunk 7]
```diff
@@ -182,3 +182,17 @@ func TestTagTuple(t *testing.T) {
...
+	mockStub := NewMockStubWithRegisterNode("substra", scc)
+	registerItem(t, *mockStub, "traintuple")
+	mockStub.MockTransactionStart("42")
+	db := NewLedgerDB(mockStub)
+	traintupleToDone(t, db, traintupleKey)
+	outTrain, err := queryTraintuple(db, keyToArgs(traintupleKey))
+	assert.NoError(t, err)
+	outPerm, err := queryModelPermissions(db, keyToArgs(outTrain.OutModel.Hash))
+	assert.NoError(t, err)
+	assert.NotZero(t, outPerm)
```
Comment (AurelienGasser) [inline chaincode/tuple_test.go:197]: IMO it would be worth adding a test that retrieves/validates permissions for a head out-model here, since (if my understanding is correct) the `queryModelPermissions` function's primary purpose is to make sure we don't try to process head out-models that don't belong to our worker.

**Review by AurelienGasser — CHANGES_REQUESTED** [review]
Great work! There are a few things worth improving IMO (nothing big), see comments below.

File: chaincode/tuple.go (hunk 2 above)
Comment (thibaultrobert) [reply chaincode/tuple.go]: Yes indeed. To you think I should removed it or used it.
On one side it will not change the number of line but on the bright side this index name will not be copied for each creation (so 4 time in the actual base code)

File: chaincode/tuple.go (hunk 4 above)
Comment (thibaultrobert) [reply chaincode/tuple.go:172]: Further more I think we should use an `InputHash` here not an `inputKey` since it's indeed a hash. @Kelvin-M do you think we could update the backend in that direction?
I'll add `modelHash` for now.

File: chaincode/tuple.go (hunk 5 above)
Comment (thibaultrobert) [reply chaincode/tuple.go]: Yep and done.

File: chaincode/tuple.go (hunk 6 above)
Comment (thibaultrobert) [reply chaincode/tuple.go]: Added

File: chaincode/tuple_test.go (hunk 7 above)
Comment (thibaultrobert) [reply chaincode/tuple_test.go:197]: For now yes and it should be extended to all models in near future. I will update the test.

File: chaincode/tuple.go (hunk 4 above)
Comment (Kelvin-M) [reply chaincode/tuple.go:172]: Yes we can, but the modification must be done also for all object to  have the same usage. So let's do it later because it will be too long to verify that everything works :)

**Comment by thibaultrobert** [comment]
Sorry for the noise, I push the wrong commit to the wrong branch.

File: chaincode/tuple.go (hunk 2 above)
Comment (thibaultrobert) [reply chaincode/tuple.go]: Done

File: chaincode/tuple.go (hunk 3 above)
Comment (thibaultrobert) [reply chaincode/tuple.go:166]: Done

File: chaincode/tuple.go (hunk 4 above)
Comment (thibaultrobert) [reply chaincode/tuple.go:172]: I was speaking about this smart contract in particular since `queryModelPermissions` returns the Permission of the Model matching the `Hash` in the arguments.

File: chaincode/tuple_test.go (hunk 7 above)
Comment (thibaultrobert) [reply chaincode/tuple_test.go:197]: Done in f3f4c7b53ea1114ad551505c2add5724a2dbfe7b.

**Review by AurelienGasser — APPROVED** [review]
